### PR TITLE
[GraphView]Add popup menu[gramps50]

### PR DESCRIPTION
**Pull request by vantu5z**
Add popup menu for node (on right mouse click).
Simplify add spouse or child to person which doesn't have family.
For node:
`- edit`
`- add new partner`
`- set as home person`

For family:
`- just open editor`

ToDo (in progress, testing and feedback needed):
After graph rebuild scroll to person which have been edited, and not to active person.

### CODE REVIEW

**Comment by fxtmtrsine** [comment]
> Add popup menu for node (on right mouse click).

@vantu5z Hope you don't mind some feed if it's not too early? No popup menu here when I click on a node (person), only when I get click on a family marriage year (node) do I get a family screen.

Love how your improving this. Thanks 👍

**Comment by SNoiraud** [comment]
If you add a popup menu, it could be interesting to add the possibility to set this node as a home person.
I have already done that in master :  see the PR #446

I get the following error when I click with the right button : 
2018-03-01 08:40:03.807: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/serge/.gramps/gramps50/plugins/GraphView/graphview.py", line 953, in select_node
    self.node_menu(node_class, handle, event)
  File "/home/serge/.gramps/gramps50/plugins/GraphView/graphview.py", line 981, in node_menu
    menu.popup(None, None, None, None, event.button, event.time)
  File "/usr/lib/python3/dist-packages/gi/overrides/Gtk.py", line 1551, in popup
    self.popup_for_device(None, parent_menu_shell, parent_menu_item, func, data, button, activate_time)
TypeError: Must be number, not EventButton

**Comment by PQYPLZXHGF** [comment]
No popup menu here also for rightclik on person or family.

> 10550: ERROR: grampsapp.py: line 143: Unhandled exception
> Traceback (most recent call last):
>   File "/home/PQYPLZXHGF/.gramps/gramps50/plugins/GraphView/graphview.py", line 953, in select_node
>     self.node_menu(node_class, handle, event)
>   File "/home/PQYPLZXHGF/.gramps/gramps50/plugins/GraphView/graphview.py", line 984, in node_menu
>     menu.popup(None, None, None, None, event.button, event.time)
>   File "/usr/lib/python3/dist-packages/gi/overrides/Gtk.py", line 1540, in popup
>     self.popup_for_device(None, parent_menu_shell, parent_menu_item, func, data, button, activate_time)
> TypeError: Must be number, not EventButton
>

**Comment by vantu5z** [comment]
@SNoiraud , don't know why you get this exception, for me it works well.
I can make like this:
```
menu.popup(None, None, None, None, event.button, event.time)
to
menu.popup(None, None, None, None, 3, event.time)
or
menu.popup(None, None, None, None, event.get_button()[1], event.time)
```

**Comment by SNoiraud** [comment]
I don't understand why we have these Gtk errors. The menu popups are created correctly for me.
I think there is a subtility somewhere...

**Comment by PQYPLZXHGF** [comment]
Not sure if anthing to do wth it, I'm using ubuntu 16.04 that has gtk 3.18.9

@vantu5z @SNoiraud @fxtmtrsine What versions are you using?  Remeber that Gramps supports gtk 3.10 or greater.

**Comment by vantu5z** [comment]
I use Arch with gtk3 3.22.28-1.

**Comment by SNoiraud** [comment]
One remark on the second entry in the meny : 
Add spouse for a women is not a good idea.
either we set Spouse/Husband or Partner or something better.
I think there was a discussion about that some time ago

**Comment by SNoiraud** [comment]
I use : 
 python    : 3.5.3
 gramps    : 5.0.0-beta2-e2426b979
 gtk++     : 3.22.11
 pygobject : 3.22.0
 pango     : 1.40.4
 cairo     : 1.14.8
 pycairo   : 1.10.0
 osmgpsmap : 1.0
 GExiv2    : 0.10
 ICU       : 57.1
 PyICU     : 1.9.5
 o.s.      : linux

**Comment by SNoiraud** [comment]
When I said very strange, this is because I wrote the popup menu the same way in geography and I have not this problem.

**Comment by vantu5z** [comment]
find that popup() is:

> Deprecated since version 3.22: Please use Gtk.Menu.popup_at_widget(), Gtk.Menu.popup_at_pointer(). or Gtk.Menu.popup_at_rect() instead

**Comment by vantu5z** [comment]
`Remeber that Gramps supports gtk 3.10 or greater.`
So we don't use new methods after gtk 3.10?

**Comment by vantu5z** [comment]
>  Add spouse for a women is not a good idea.
Why not, translators adopt it.
But think we can use 'partner' instead.

**Comment by Nick-Hall** [comment]
> So we don't use new methods after gtk 3.10?

Correct.

**Comment by vantu5z** [comment]
>     So we don't use new methods after gtk 3.10?
> Correct.

Ok. Will use: menu.popup(None, None, None, None, event.get_button()[1], event.time)

**Comment by vantu5z** [comment]
Try to focus on person (not active) after editing or add family and graph rebuilding.

**Comment by vantu5z** [comment]
As solution I see only change active person.
Any advice will good :)

**Comment by fxtmtrsine** [comment]
@PQYPLZXHGF ubuntu 14.04 with gtk 3.10.8  looks like I have the oldest!

@vantu5z With the latest version still get no popup menu here for the person; only clicking on a family marriage works as before. so no change for me :(

**Comment by vantu5z** [comment]
@fxtmtrsine Try now, rewrite menu creation by gramps widgets.

**Comment by sam-m888** [comment]
Tested it myself and get no pop up context menu(right click) for person nodes using the MS-Windows Gramps AIO 5.0 beta 1 (32 or 64bit)[gtk 3.18.9 & using example.gramps ] , only family nodes show the Family Editor.

A few times I saw a quick flash & shadow that could be indicative of context menu.

Also going from active person to active person with a left click the Gramps program title shows the word (unresponsive ) about 8 seconds then the following was shown in the terminal:

> Warning: History: Signal recursion blocked. Signals was : active-changed
>         from: file: C:\Program Files (x86)\GrampsAIO32-5.0.0-beta1\gramps\gui\displaystate.py
>               line: 158
>               func: push
> 

Tried it on Ubuntu 16.04 (gtk 3.18.9) same result for the context menu but not the unresponsive navigation!

**Comment by sam-m888** [comment]
> make 'self.menu', maybe problem in this same in editperson and pedigree

Gramps AIO 5.0.0 beta 1 Win 10  shows the context menu 👍 ,  but remains unresponsive as mentioned earlier.

graphview-add-context-menu

**Comment by sam-m888** [comment]
Ubuntu 16.04 works with no issue

graphview-add-context-menu-ubuntu

**Comment by bpisoj** [comment]
@sam-m888
For me GrampsAIO64-5.0.0-beta1 works with latest modification
https://github.com/vantu5z/addons-source/blob/c46b5b0e4e39f7d84499f0ea83c1ec96038fb89d/GraphView/graphview.py

**Comment by sam-m888** [comment]
@bpisoj Yes works for me with the 64bit Gramps AIO build 3, both menu and
the chart is responsive.

@prculley My issue occurs with the 32bit Gramps AIO build 2, as shown in
the error message in the comment
https://github.com/gramps-project/addons-source/pull/120#issuecomment-370046412

@bpisoj  @prculley I'll attempt to be a little more clearer when it comes
to the AIO bundles as they are built by two different people, Thank you to
both of you :)

**Comment by fxtmtrsine** [comment]
> @fxtmtrsine Can you test with ubuntu 14.04 please?

All good with the latest version 👍 

@vantu5z Would it make sense to be able to change the line types (curves, ortho, none) by right clicking?

**Comment by vantu5z** [comment]
`line types (curves, ortho, none) by right clicking?`
@fxtmtrsine  Is this so useful? If so, it can be implement to right click on background.

**Comment by vantu5z** [comment]
> Also going from active person to active person with a left click the Gramps program title shows the word (unresponsive ) for about 8 seconds

@sam-m888 Is this came from this PR or occur before it?

**Comment by sam-m888** [comment]
@vantu5z This PR only shows the issue and is not the cause.  eg: Only the
Gramps AIO 32 bit build 2 seems to have an issue.

**Comment by fxtmtrsine** [comment]
> line types (curves, ortho, none) by right clicking?
> @fxtmtrsine Is this so useful? If so, it can be implemented to right click on background.

@vantu5z Personally yes. What I'm thinking of is now that you have the popup menu; that most of the config options that are hidden can be available, you can see this at work in the TimelinePedigreeView addons popup menu which has no active config option icon. What do you think?

**Comment by bpisoj** [comment]
@sam-m888
I don't see problem with GrampsAIO32 build 2

**Comment by prculley** [comment]
@sam-m888 I'm afraid I don't see any issues on the 32bit AIO either.

**Comment by sam-m888** [comment]
@bpisoj @prculley Going on the fact you both tested and could not replicate
I reloaded a clean Win 10 (32bit) Virtual Machine and downloaded the Gramps
AIO32 build 2 installer and tested again and it worked first go! Well that
was frustrating!

The only thing I can think of is some interaction between the pycache files
from the Gramps AIO64 version that is also installed on that dodgy VM!

Going back to the dodgy VM that shows the issues I'll paste the final
output before I delete the VM:

GrampsAIO32 build 2
Graph view (setting: 1 ancestors / 1 descendants)

(gramps.exe:7696): Gtk-WARNING **: GtkEntry - did not receive
focus-out-event. If you
connect a handler to this signal, it must return
FALSE so the entry gets the event as well

(gramps.exe:7696): Gtk-WARNING **: GtkEntry - did not receive
focus-out-event. If you
connect a handler to this signal, it must return
FALSE so the entry gets the event as well

(gramps.exe:7696): Gdk-CRITICAL **: gdk_device_get_source: assertion
'GDK_IS_DEVICE (device)' failed

**Comment by prculley** [comment]
@sam-m888 You really should not have had to redo your VM; it should be sufficient to delete the AIO installation directory and make a new one.  I really hate the `(gramps.exe:7696):` type of warnings and errors; I have no idea what the number means...  I get the same thing from Python sometimes.

**Comment by fxtmtrsine** [comment]
> move lines type setting to popup menu

@vantu5z 👍   Your the best. 

Does it make sense to have two separate popup menus, why not join them?

**Comment by vantu5z** [comment]
> Does it make sense to have two separate popup menus, why not join them?

@fxtmtrsine How do you see this? One contain elements for nodes, another for global settings.

> What I'm thinking of is now that you have the popup menu; that most of the config options that are hidden can be available

Yes, I work on this.

**Comment by vantu5z** [comment]
Add <Edit tags list> for person. Can someone review saving procedure?

**Comment by vantu5z** [comment]
> Graph view (setting: 1 ancestors / 1 descendants)

(Gramps.py:11429): Gdk-CRITICAL **: gdk_device_get_source: assertion 'GDK_IS_DEVICE (device)' failed

@sam-m888 I can get this once, but can't reproduce.

**Comment by fxtmtrsine** [comment]
> Add (Edit tags list) for person. Can someone review saving procedure?

Yes worked for me. 👍 

> How do you see this? One contain elements for nodes, another for global settings.

Yes join the popmenus together like in [timeline](https://gramps-project.org/wiki/index.php?title=TimelinePedigreeView) addon ([like i mentioned earluier](https://github.com/gramps-project/addons-source/pull/120#issuecomment-370202371) install and try it 👍 ) it uses one menu that shows the edit bit at top divided by a line from the config stuff and also has a link to the help page so for graphviw it would be [this page](https://gramps-project.org/wiki/index.php?title=Graph_View)

This means tht you can also use menu on marriage ovals **Edit & Edit Tags** instead of directly opening marraige.  heres my crappy image 

menu like timeline

Question I look at option for **path color** and change color but see no change, which path does it mean?

**Comment by SNoiraud** [comment]
> Does it make sense to have two separate popup menus, why not join them?

I am not sure the popup menu for the lines type is a good idea.
Once you choosed one, you never change that in the future.

I really don't like the popup menu on the node because I can't edit directly the person like for the family.
So to have one menu for all is a bad idea for me.

In the worst case, one menu for the node, one for the family and one for extra options (bad idea).

**Comment by vantu5z** [comment]
> which path does it mean?

@fxtmtrsine This is the path to the home person from active.

> In the worst case, one menu for the node, one for the family (for adding tags) and one for extra options (bad idea)

@SNoiraud I don't understand you. 
- Don't you like any popup at all?
- Don't you like only extra options popup? 
- Or separated menus (with items for actions with object) can be?

For me I don't like one menu for all. It looks overloaded.
About extra menu (click on background): I don't see any enhancement in it, we always can press config button and set all we want. And as @SNoiraud said "Once you choosed one, you never change that in the future" - I agree with him.

So, I will add menu for family (edit and tags). And wait for feedback and ideas about extra menu.

**Comment by johnbickmore** [comment]
> For me I don't like one menu for all. It looks overloaded.

@vantu5z You're a bit late to the game with Gramps then! Have a look at the context menus for almost all the Chart views(I believe the Quilt chart is the odd duck and really needs a menu also!). Ideally carrying through the same type of menu to Graph View will at least make it consistent, as personally I've never liked how it differs from the rest in the way that it works at least with your menu it not yet another Gramps mystery tour of guess what happens when I press this!  And the menu on family works for me, good stuff. Consider what the rest of the program is like and not just the Graph view!

As for the options on the menu like the timeline addon I'd say add all of them, and that config button is not very obvious for a lot of us!

**Comment by vantu5z** [comment]
@johnbickmore Thanks for your feedback!
I'll try to make it similar to other popups we already have.

**Comment by SNoiraud** [comment]
>    Don't you like any popup at all?

Yes if you have a small number of entries. It seems good to me with three entries for the node
I said that because I use this view for years and only clicking on the node called the editor. I need now to choose edit in the menu.
For example, the timeline popup menu is too complex and have too many entries for me.

>   Don't you like only extra options popup?

If we need to use them at any moment, Yes. Actually, the path color could be in the background popup menu and the line type too, but I never use them. So where is the improvement ? This add complexity to the program.

>    Or separated menus (with items for actions with object) can be?

Yes. the popup menu for the nodes and families are fine for me. Your last menu for families with add tags is good for me.

**Comment by vantu5z** [comment]
@SNoiraud Do you have any issue with focus on edited person or family?
If we will have twice populating (rebuilding) for some reason, it can be focused on active person instead of edited. This is not big issue but any way.

**Comment by SNoiraud** [comment]
@johnbickmore
> I believe the Quilt chart is the odd duck and really needs a menu also!

If you want such a thing, create a feature request for that.
What kind of menu entries do you want ?
For this view we have only four options : 
center the person, the path color, the path transparency and the selected person color
Once we have choosed the values for our needs, we never change that, So, the menu seems useless
perhaps you nedd something else ?

**Comment by SNoiraud** [comment]
> Do you have any issue with focus on edited person or family?

For the moment, I saw no differences, and the focus seems correct for me.

**Comment by vantu5z** [comment]
How would Bender said: "We will make our own with blackjack and something else"  ^_^
It's joke. But we try make it better.

> I saw no differences, and the focus seems correct for me

@SNoiraud Before PR it focused on active person, so after editing person by right click you move to active person, for now behavior is differs.

**Comment by vantu5z** [comment]
@SNoiraud I will try to add open editor by double click as more commonly used action.

**Comment by SNoiraud** [comment]
Do you means we focus on the edited person instead of the active one ? This is what happens to me.
Why not.

**Comment by SNoiraud** [comment]
It depends on the modifications you are doing. In some cases, I go to the active person and in some other cases, to the edited person. Yes, this is a problem. We must always do the same thing.

**Comment by sam-m888** [comment]
> You really should not have had to redo your VM

@prculley That's ok, I needed to build one specifically for 32bit on Win 10 instead of only testing against 64bit.

> I can get this once, but can't reproduce.

@vantu5z I can no longer reproduce as I reset that VM.

**Comment by johnbickmore** [comment]
> What kind of menu entries do you want?

@SNoiraud As mentioned in my earlier [comment](https://github.com/gramps-project/addons-source/pull/120#issuecomment-370775927), a menu consistent with the other Gramps views this would allow using the menu to goto to the children or ancestors etc more easily like the other charts, But I agree for this chart just going to another person may not mean that you want to change the selected person line for the chart(this would be different from the other chart views!) only that you want to navigate up or down a generation to see where that highlighted line leads you. not sure if I'm clear on this point.
And possibly a small toolbar bar like Graphview has, this would allow you to zoom in and out and get a better overview.  Another option would be to be able to a the quilt gramplet in the sidedock an have it be zoomed out so you can use it to navigate around the larger view.

> For this view we have only four options :
> center the person, the path color, the path transparency and the selected person color
> Once we have choosed the values for our needs, we never change that, So, the menu seems useless
> perhaps you need something else ?

I agreed with the exception of the "center the person" option which if you turn it off makes using the chart erratic at best give it a try an see what I mean, especially when you drag the chart around and accidentally touch a person record in an attempt to drag the chart, I'd prefer that the chart centered on the mouse or if not possible the new position of the person be animated to draw the eye to it like in graph view.

One other suggestion while I'm at it is that when you go to choose a family to edit you unfortunately only see a line of "f"'s, 

lots-of-f s-let-guess-which-one-i-want


it would be great if when you hovered over one of them it showed a tooltip for which couples family record it is for. Bonus for something similar for the person. the other charts do it, have a look? By the way what do the black squares and circles mean? How about a tooltip for them also?

> If you want such a thing, create a feature request for that.

@SNoiraud Great work you've done love the geography views 🥇  I'm sorry for the long rant. didin't intend this to be a mini review and thanks for the suggestion to create a feature request but I'll pass on that as it would mean getting yet another account, I'd rather the project turn on Github issues like everyone else! If someone can raise a feature request I'd appreciate it 👍

**Comment by vantu5z** [comment]
@johnbickmore Make menu for person like at fanchart.

@SNoiraud Try it, I found it useful. Helps  very with navigation to children or siblings.

Have some idea. If person in menu lists have not present in current graph - I think we should not just ignor that, but show some dialog to make it active and rebuild graph. What do you think about this.

Add shaking person after editing. That helps find them because in some cases it placed not at the screen center.

**Comment by SNoiraud** [comment]
I don't like that, but it works.

> I think we should not just ignor that, but show some dialog to make it active and rebuild graph.
YES, it should be better.

**Comment by vantu5z** [comment]
> I don't like that

@SNoiraud Why not?
You don't need to follow the lines by eyes, just look to the list and click, and get animation to selected person. And other useful stuff as `add partner`, `add parents`, `add_tags`, `goto siblings` that can be far from each other.

Now I try to perform double click to edit person or family without popup.
So using popup is stay only if you want to do something else.

**Comment by SNoiraud** [comment]
Sound good to me.
Tested all cases.
The only remaining problem is go to person not already displayed in the current graph.

**Comment by vantu5z** [comment]
> The only remaining problem is go to person not already displayed in the current graph.

@SNoiraud Try it now :)

**Comment by SNoiraud** [comment]
Yes it works.

I had the following problems when i modify the number of descendants: 

07075: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/serge/.gramps/gramps50/plugins/GraphView/graphview.py", line 2759, in stop_shake_animation
    if (self.counter[item.title] < self.max_count) and (not stoped):
KeyError: 'b3d9dfcc99644708fd6'

**Comment by vantu5z** [comment]
@SNoiraud What do you about family menu, should we add something else?

**Comment by SNoiraud** [comment]
No more problems.
OK for children submenu.

**Comment by sam-m888** [comment]
@vantu5z Looks good, haven't found any issue added from this PR. Is it almost ready to merge?

**Comment by vantu5z** [comment]
@sam-m888 It is not ready yet. 
Extra menu (on background) is not implemented properly.

**Comment by vantu5z** [comment]
Add items from config to extra menu (on background).
Think that color and animation settings should not be in this menu.
So config dialog should exist.

Should we add something else?

File: GraphView/graphview.py [hunk 1]
```diff
@@ -1023,6 +1023,59 @@ def background_menu(self, event):
         self.menu = Gtk.Menu()
         self.menu.set_reserve_toggle_size(False)
 
+        menu_item = Gtk.CheckMenuItem(_('Show images'))
```
Comment (Nick-Hall) [inline GraphView/graphview.py]: I thought we had decided to move away from putting configuration settings in the context menus

**Comment by sam-m888** [comment]
@vantu5z Tested ok, so far 👍 

> Should we add something else?

Possibly like [mentioned](https://github.com/gramps-project/addons-source/pull/120#issuecomment-370641262) by @fxtmtrsine  also add an identical help entry to each of the three menus called "About Graph View" that leads to the url:  https://gramps-project.org/wiki/index.php?title=Graph_View  like shown in Timeline pedigree and htreeview eg:

2018-03-12 12_22_38-timelinepedigreeview-addon-example-50 png png image 991 x 801 pixels

Apart from that I'd say this is finished as is. 👍 

Screenshots for everyone of the three menus

config-menu-graphview-50
edit-family-menu-graphview-50
edit-person-menu-graphview-50

**Comment by sam-m888** [comment]
@vantu5z Nicely done 🥇 When you're ready for this to be merged remove the "Work in Progress" label.

Screenshots of the three menus:

Config menu:
config-menu-graphview-50

Edit Family menu:
edit-family-menu-graphview-50

Edit Person menu:
edit-person-menu-graphview-50

**Comment by vantu5z** [comment]
Yes, I think it can be merged.

**Comment by sam-m888** [comment]
Thank you

**Pull request merged**